Add WHM spell: Arise

Omnione opened the pull request:
Addition of **Arise** WHM spell level 99

A few changes have been made in relation to double weakness tracking in the PR,
Previously the tracking was done via checking if the player had weakness on while dead,
Retail does not do this, if your are weakend and die, the weakness effect is removed,
This will now change the behavior of weakness so it gets removed upon death if the effect is currently active,
weakness will then be re applied with the new values (previously weakness was being ticked down while dead).
Now it will be traked in the same manner as MijinGakure with a local variable.

All the tracking is done in core in charentity.

-- Added the script for Arise
Arise tracking and removal of the variable used is handled in charentity so reraise can be applied on raise, 
But the variable is set in lua.

The timer for weakness is also reduced upon raise from arise from 5 mins to 3 mins.

All exp loss values and hp% returned values as per BG wiki.

Video for testing results:
https://youtu.be/peGSB4h9fP8

Related issue: WHM99 Spell: Arise - missing SQL entry and Lua script #1306

Would appreciate some testing from others to confirm changes are ok.
 
**_I affirm:_**
- [x] I have paid attention to this example and will edit again if need be to not break the formatting, or I will be ignored
- [x] that I agree to LandSandBoat's [Limited Contributor License Agreement](https://github.com/LandSandBoat/server/blob/base/.github/CONTRIBUTOR_AGREEMENT.md), as written on this date
- [x] that I have **read the [Contributing Guide](https://github.com/LandSandBoat/server/blob/base/CONTRIBUTING.md)**
- [x] that I've _**tested my code and things my code changed**_ since the last commit in the PR, and will test after any later commits

Omnione committed:
Arise addition

Xaver-DaRed commented on scripts/globals/spells/white/arise.lua:
```diff
@@ -0,0 +1,32 @@
+            target:setLocalVar("Raise", 1)
+            target:entityAnimationPacket("sp00")
+            target:addHP(target:getMaxHP())
+            target:addMP(target:getMaxMP())
```
Wouldnt his block also work with Prishe, the trust?
Maybe a check for zone, or battlefield is in order

Omnione replied:
Shouldn’t be a problem, trust disappear as soon as they die, never giving a player a chance to raise them, this has been in every raise script previously and was added in this one for consistency

TeoTwawki replied:
somewhere in the future we should consider putting some kind of hook to the raise function to perform the work instead of duplicate scripting but even if trusts didn't despawn non players also require a target flag to even have raise cast on them and to my knowledge only that specific version of Prishe has it set out of all our entities in the database.

Omnione committed:
[Arise] Adjusted tracking

Added new variables for charentity,
m_weaknessLvl: to track weakness levels.
m_hasArise: bool to track reraise application after Arise

Changed weakness.lua to only check if weakness effect power is > 1 rather than == 2, this is to accommodate the new variable m_weaknessLvl to be used.

arise.lua now sends a value of 5 for raise level
added a new level of raise to charentity of 5 to handle arise.

TeoTwawki commented on src/map/entities/charentity.cpp:
```diff
@@ -1587,6 +1596,13 @@ void CCharEntity::OnRaise()
             hpReturned             = (uint16)GetMaxHP();
             ratioReturned          = ((GetMLevel() <= 50) ? 0.50f : 0.90f) * (1 - map_config.exp_retain);
         }
+        else if (m_hasRaise == 5)
```
seems t4 and t4 have identical code here, but I can't locate where in the project we actually set t4

Omnione replied:
yeah that baffled me also, but i did not want to touch it as it was obviously put there for a reason

Omnione replied:
unless this was for arise but never commented as such

TeoTwawki replied:
after trying to dig around in the project..I think that may be the case. Gonna see what git blame says it came in with

TeoTwawki replied:
@claywar git blame points its finger at you, was this intended for Arise ?
https://github.com/LandSandBoat/server/commit/053211ee5af7db48830776adaee9ec4a7ee7aab1#diff-81491448fbf9d158d4cc9e7df3822956203b840795b53b030ee1be7f250dd7cfR1575

Omnione replied:
Ah, Clay added it for Reraise 4 from the job point additions

TeoTwawki replied:
There is a reraise 4?  /noob

Omnione replied:
from BG
Grants an [Arise](https://www.bg-wiki.com/ffxi/Arise) effect upon KO that restores 100% HP and 90% XP lost and grants 3 minute weakness.

Omnione replied:
so yup, can use it

TeoTwawki replied:
I have to keep manually refreshing to see your messages, github being garbage today

Omnione replied:
yeah same

Omnione replied:
ok, ill change this to use 4 instead, and that means reraise 4 can be plugged in also

claywar replied:
I believe that was meant for Arise, so yes.

TeoTwawki replied:
I guess Arise is basically raise4 to match that reraise4, but with the added benefit of working for the dead person _twice_ (since your whm isn't dead on self casting, they only get the reraise half of it)

Omnione replied:
yeah, reraise with reraise

TeoTwawki replied:
Yo dawg, I heard you like reraise...

Omnione committed:
Changed the number that was being sent to charentity.cpp for OnRaise from 5 back to 4 as this was intented for Arise when added.
